sklearn/preprocessing/_encoders.py

```diff
@@ -724,11 +727,28 @@ def get_feature_names_out(self, input_features=None):
 
         feature_names = []
         for i in range(len(cats)):
-            names = [input_features[i] + "_" + str(t) for t in cats[i]]
+            names = [str(input_features[i]) + "_" + str(t) for t in cats[i]]
             if self.drop_idx_ is not None and self.drop_idx_[i] is not None:
                 names.pop(self.drop_idx_[i])
             feature_names.extend(names)
-        return np.asarray(feature_names, dtype=object)
+        return np.asarray(
+            self._ensure_unique_feature_names(feature_names), dtype=object
+        )
+
+    def _ensure_unique_feature_names(self, feature_names):
+        name_counts = defaultdict(int)
+        unique_feature_names = []
+        for name in feature_names:
+            if name in name_counts:
+                original_name = name
+                suffix_count = name_counts[original_name]
+                name = original_name + "__" + str(suffix_count)
+                while name in name_counts:
+                    suffix_count += 1
+                    name = original_name + "__" + str(suffix_count)
```

Comment:
As with any automatic solution, there will be weird edge cases. For example, with this PR:

```python
from sklearn.preprocessing import OneHotEncoder
import pandas as pd

df = pd.DataFrame([["None"], [None], ["None__1"], ["None__2"]])
enc = OneHotEncoder().fit(df)

print(enc.get_feature_names_out())
# ['x0_None' 'x0_None__1' 'x0_None__2' 'x0_None__3']
```

It becomes difficult to figure out the correspondence between the categories and the feature names.

If we go with this, we would need to explain the suffixing behavior and how the feature names match up with the categories.

Reply:
Totally agree about the edge cases, it'd be very difficult to figure out where it came from. 

The question that comes to me which option is more preferred
- (more forgiving) this implementation as a default solution (always working, but sometimes confusing) + ability to change the naming function
- (more explicit) raising in case of non-unique feature names + ability to change the naming function.

_It also might make sense to change the default function from using `str` to using `repr` in some future versions._

---

Which kind of interface did you have for that naming function?
One possible implementation is `(feature: str, value: str) -> str`.

Reply:
It would be `(feature: Any, category: Any) -> str`. Sometimes the category can be ints or `None`. And the `feature` can be ints.

The nice thing about the callable is that we can have options denoted as strings. Let's say the parameter is called `feature_name_combiner` (for the lack of a better name). It can accept a callable to allow users to combine the feature and category themselves. In the  future, we can add an option like `feature_name_combiner="auto_suffix"` to do auto suffixing,  which would be a more forgiving setting.

Reply:
> It would be (feature: Any, category: Any) -> str. Sometimes the category can be ints or None. And the feature can be ints.

You're right. I've figured I had written the wrong signature but didn't have access to my computer after that. Hope that didn't confuse you too much.

> The nice thing about the callable is that we can have options denoted as strings. Let's say the parameter is called feature_name_combiner (for the lack of a better name). It can accept a callable to allow users to combine the feature and category themselves. In the future, we can add an option like feature_name_combiner="auto_suffix" to do auto suffixing, which would be a more forgiving setting.

That's a good point, that interface would make the configuration easier, and is already present in many transformers.